### DISCUSSION TITLE
The best way to change reducer's environment dynamically

### DISCUSSION THREAD

**Discussion by satanworker** [pointfreeco/swift-composable-architecture#759] [Q&A] [answered]
I'm not sure if that's the proper way to do, but I have an app which would work with multiple cloud providers like iCloud, Dropbod, Gdrive and reducer logic will be pretty much the same but I don't know which provider will be used until user input 
so what is the best way to change environment?
if I do something simple like this
```
struct UserEnvironment {
    var cloudClient: CloudClient
}
```

<img width="537" alt="Screen Shot 2021-09-01 at 22 12 43" src="https://user-images.githubusercontent.com/11026240/131730404-d6edcb9d-971d-4e90-b762-b19d23d34a28.png">
<img width="1004" alt="Screen Shot 2021-09-01 at 22 09 45" src="https://user-images.githubusercontent.com/11026240/131730441-a634c365-e9eb-439d-9e16-93a0f998cb93.png">

I still get an error 
<img width="1037" alt="Screen Shot 2021-09-01 at 22 13 08" src="https://user-images.githubusercontent.com/11026240/131730456-25478b11-5130-42c3-b21c-a31eea5ceb78.png">
even though XCode see userEnvironment.cloudClient as variable

**Comment by satanworker** [comment]
sorry for posting the same screenshot twice

**Comment by tgrapperon** [comment, marked as answer]
I would recommend not mutating the environment in the reducer, like you do for `credentialStorage`. You don't have an error at this line because `credentialStorage` is most likely a class.

Ideally, your reducer only mutates the state and produce effects. These effects can involve the environment, of course, and you can also produce effects that will mutate the environment.

In your case, I would wrap your client inside another "meta" client that is configurable. When you receive `createNewPassword`, you emit some effect configuring the "meta" client. You can even chain directly with `encryptNewPassword` without getting back into the reducer. In other words, try to move the two lines involving the environment in an effect that will either:
- Produce another effect that will produce the `encryptNewPassword` effect when getting back into the reducer (the recommended way)
- Chain directly the result with the `encryptNewPassword` effect without producing an intermediary effect (less optimal, as less testable and bearing some logic)

Your meta `cloudClient` can be a class, so you can mutate things (like the inner `cloudClient`), but this is something that happens in the outside world (i.e., through an effect), not in the reducer. The reason is that something can change the `credentialStorage` or the `cloudClient` outside the reducer's scope, producing untestable and unexpected side effects. You should ideally also watch these changes if they can happen, and produce according effects to update the state.
